Plutarch/Prelude.hs

```diff
@@ -13,17 +14,143 @@ module Plutarch.Prelude (
   (#),
   plet,
   pinl,
-  pcon,
-  pmatch,
   pto,
   pfix,
   Type,
   S,
   PType,
+  PlutusType,
+  PCon (pcon),
+  PMatch (pmatch),
+
+  -- * Integers and integer utilities
+  PInteger,
+  PIntegral (pdiv, pmod, pquot, prem),
+
+  -- * Rational numbers and utilities
+  PRational,
+  pnumerator,
+  pdenominator,
+  pfromInteger,
+  pround,
+
+  -- * Booleans and boolean functions
+  PBool (..),
+  PEq ((#==)),
+  POrd ((#<=), (#<)),
+  pif,
+  pnot,
+  (#&&),
+  (#||),
+
+  -- * Bytestrings and bytestring utilities
+  PByteString,
+  phexByteStr,
+  pconsBS,
+  psliceBS,
+  plengthBS,
+  pindexBS,
+
+  -- * String and string utilities
+  PString,
+  pencodeUtf8,
+  pdecodeUtf8,
+
+  -- * Unit type and utilities
+  PUnit (..),
+
+  -- * Common list typeclass and utilities
+  PListLike (PElemConstraint, pelimList, pcons, pnil, phead, ptail, pnull),
+  PIsListLike,
+  plistEquals,
+  pelem,
+  plength,
+  ptryIndex,
+  pdrop,
+  psingleton,
+  pconcat,
+  pzipWith,
+  pzipWith',
+  pzip,
+  pmap,
+  pfilter,
+  precList,
+  pfoldr,
+  pfoldrLazy,
+  pfoldl,
+  pall,
+  pany,
+
+  -- * Scott encoded list type
+  PList (..),
+
+  -- * Scott encoded maybe type and utilities
+  PMaybe (..),
+
+  -- * Scott encoded either type and utilities
+  PEither (..),
+
+  -- * Scott encoded pair type and utilities
+  PPair (..),
+
+  -- * Builtin types and utilities
+  PData (..),
+  pfstBuiltin,
+  psndBuiltin,
+  PBuiltinPair,
+  PBuiltinList (..),
+  PIsData (pfromData, pdata),
+  PAsData,
+  ppairDataBuiltin,
+  PBuiltinMap,
+
+  -- * DataRepr and related functions
+  PDataFields,
```

Comment:
I was unsure whether or not the DataRepr stuff should be added to prelude. I think they should be - but I don't know if these exports are all that's needed or if there are more commonly used functions that should also be exported here.

Reply:
The issue is you'd have to add all the other stuff, like generics, too. Maybe it's worth it.

Reply:
Which imports, in particular, are needed in association with this? You don't need `GHC.Generics.Generic` to *use* data repr stuff right? Just to derive it for your own type.

Reply:
We don't need to export `PDataFields` if you do not intend to allow defining data types with just the prelude.

Reply:
Removed `PDataFields` export.